doc/tablator.1
.TH TABLATOR "1" "October 2017" "tablator 4.0.1" "User Commands"
.SH NAME
tablator - The Universal Table Translator for Astronomy
.SH SYNOPSIS
tablator [options] [INPUT] [OUTPUT]
.SH DESCRIPTION
Tablator is a utility for converting tables in astronomy into
different formats.  Tablator can read and write tables in fits, ipac
table, hdf5, votable, json, json5, csv, and tsv, and can write html,
SQL commands, and SQLite databases.  The emphasis is on speed, so all
conversions are done in memory.
.PP
Unless passed the \fB\-\-input-format\fR option, tablator attempts to
guess the formats of the input file by looking at the file extension.
If it can not guess the format of the input file, then it will try all
of the formats it understands.
.PP
Tablator also tries to deduce the format of the output file by looking
at the file extension.  If OUTPUT is a single hyphen (-), then the result
is written to standard output.
.PP
The supported file types and their recognized extensions are
.TP
.B JSON
js, json
.TP
.B JSON5
js5, json5
.TP
.B VOTable
xml, vot, vo
.TP
.B HDF5
h5, hdf, hdf5
.TP
.B Comma Separated Values (CSV)
csv
.TP
.B Tab Separated Values (TSV)
tsv
.TP
.B FITS
fits
.TP
.B IPAC Table
tbl, no extension
.TP
.B Text
txt
.TP
.B HTML
html
.TP
.B Postgres compatible SQL
postgres
.TP
.B SQLite compatible SQL
sqlite
.TP
.B Oracle compatible SQL
oracle
.TP
.B SQLite single file database
db
.PP
Note that a file with no extension (e.g. \fBmytable\fR) is considered an IPAC table by default.
.PD
.SH OPTIONS
.PP
.TP
\fB\-\-help\fR
show a help message and exit
.TP
\fB\-\-input-format\fR=\fI\,FORMAT\/\fR
Format of the input file.  Allowed values for FORMAT are
.br
.B json
.br
.B json5
.br
.B votable
.br
.B hdf5
.br
.B csv
.br
.B tsv
.br
.B fits
.br
.B ipac_table
.TP
\fB\-\-output-format\fR=\fI\,FORMAT\/\fR
Format of the output file.  Allowed values are those taken by the \fB\-\-input-format\fR option and
.br
.B text
.br
.B html
.br
.B postgres
.br
.B sqlite
.br
.B oracle
.br
.B db
.SH "EXAMPLES"
To convert an HDF5 file named mytable.hdf5 to fits
.PP
.br
.B tablator mytable.hdf5 mytable.fits
.TP
To convert a VOTable file named mystery.table to a JSON file named mystery_json.table.
.PP
.br
.B tablator --input-format=votable --output-format=json \e
.br
.RS
.B mystery.table mystery_json.table
.RE
.TP
To convert a VOTable to an SQLite database, select only ra, dec, and object, sort by object, and write a JSON file
.PP
.br
.B tablator mytable.xml mytable.db
.br
.B sqlite3 -csv -header mytable.db """select ra,dec,object from mytable order by object;""" \e
.br
.RS
.B | tablator --input-format=csv - result.json
.RE

.SH COPYRIGHT
2014-2017 California Institute of Technology



debian/changelog
tablator (4.0.1+dfsg-1) unstable; urgency=low

  * Make CSV/TSV input handle dos-style line endings
  * Add a filtering example in the manpage

 -- Walter Landry <wlandry@caltech.edu>  Fri, 13 Oct 2017 17:47:41 -0700

tablator (4.0.0+dfsg-1) unstable; urgency=low

  * Add JSON and JSON5 to content_type()
  * Fix quoting of strings for sql output
  * Fix arrays for sql output
  * Rename write_*_sql to write_sql_* and add write_sql_insert

 -- Walter Landry <wlandry@caltech.edu>  Tue, 3 Oct 2017 16:14:08 -0700

tablator (3.0.2+dfsg-2) unstable; urgency=low

  * Incorporate multithreading fix from json5_parser

 -- Walter Landry <wlandry@caltech.edu>  Mon, 14 Aug 2017 14:59:20 -0700

tablator (3.0.2+dfsg-1) unstable; urgency=low

  * Allow DOS line ending in IPAC tables

 -- Walter Landry <wlandry@caltech.edu>  Sat, 29 Jul 2017 13:09:31 -0700

tablator (3.0.1+dfsg-1) unstable; urgency=low

  * Fix escaping when writing CSV or TSV

 -- Walter Landry <wlandry@caltech.edu>  Wed, 19 Jul 2017 21:36:34 -0700

tablator (3.0.0+dfsg-1) unstable; urgency=low

  * Rename tablator::Table::write_output() to write()
  * Remove tablator::Table::output_precision.
  * Add support for pure stream reads.

 -- Walter Landry <wlandry@caltech.edu>  Sun, 16 Jul 2017 18:02:17 -0700

tablator (2.3.5+dfsg-1) unstable; urgency=low

  * Properly escape newlines when writing ipac headers

 -- Walter Landry <wlandry@caltech.edu>  Wed, 21 Jun 2017 16:16:40 -0700

tablator (2.3.4+dfsg-1) unstable; urgency=low

  * Properly write out varchar's size for Oracle and Postgres

 -- Walter Landry <wlandry@caltech.edu>  Wed, 21 Jun 2017 11:56:58 -0700

tablator (2.3.3+dfsg-1) unstable; urgency=low

  * Omit the trailing semicolon in write_create_table_sql

 -- Walter Landry <wlandry@caltech.edu>  Wed, 21 Jun 2017 11:17:47 -0700

tablator (2.3.2+dfsg-1) unstable; urgency=low

  * Only quote sql if using non-regular identifiers

 -- Walter Landry <wlandry@caltech.edu>  Wed, 21 Jun 2017 10:29:16 -0700

tablator (2.3.1+dfsg-1) unstable; urgency=low

  * Write csv/tsv nulls as empty columns
  * Interpret empty columns in IPAC_TABLE as nulls
  * Better error messages when reading a table with no columns

 -- Walter Landry <wlandry@caltech.edu>  Tue, 20 Jun 2017 14:31:55 -0700

tablator (2.3.0+dfsg-1) unstable; urgency=low

  * Allow writing CREATE TABLE and the INSERT clauses separately.

 -- Walter Landry <wlandry@caltech.edu>  Wed, 14 Jun 2017 13:29:53 -0700

tablator (2.2.0+dfsg-1) unstable; urgency=low

  * Support writing SQL text and binary SQLite databases

 -- Walter Landry <wlandry@caltech.edu>  Fri, 9 Jun 2017 21:39:25 -0700

tablator (2.1.6+dfsg-1) unstable; urgency=low

  * Run through clang-format.  No code changes.

 -- Walter Landry <wlandry@caltech.edu>  Wed, 26 Apr 2017 10:35:32 -0700

tablator (2.1.5+dfsg-1) unstable; urgency=low

  * Make html output more spiffy

 -- Walter Landry <wlandry@caltech.edu>  Wed, 22 Mar 2017 17:24:56 -0700

tablator (2.1.4+dfsg-1) unstable; urgency=low

  * Handle most attributes in links

 -- Walter Landry <wlandry@caltech.edu>  Mon, 20 Mar 2017 11:17:02 -0700

tablator (2.1.3+dfsg-1) unstable; urgency=low

  * Handle most attributes in links

 -- Walter Landry <wlandry@caltech.edu>  Mon, 20 Mar 2017 07:51:51 -0700

tablator (2.1.2+dfsg-1) unstable; urgency=low

  * Handle single quoted hyperlinks in HTML

 -- Walter Landry <wlandry@caltech.edu>  Sun, 19 Mar 2017 11:30:44 -0700

tablator (2.1.1+dfsg-1) unstable; urgency=low

  * Do not escape hyperlinks when writing HTML

 -- Walter Landry <wlandry@caltech.edu>  Sat, 18 Mar 2017 23:21:43 -0700

tablator (2.1.0+dfsg-1) unstable; urgency=low

  * Export write_type_as_ascii publicly

 -- Walter Landry <wlandry@caltech.edu>  Thu, 16 Mar 2017 13:07:09 -0700

tablator (2.0.1+dfsg-1) unstable; urgency=low

  * For VOTable output, Write the types for array-valued fields correctly

 -- Walter Landry <wlandry@caltech.edu>  Sun, 5 Mar 2017 12:00:00 -0700

tablator (2.0.0+dfsg-1) unstable; urgency=low

  * Read and write PARAM and FIELD metadata in votable and hdf5
  * Now requires C++ 14

 -- Walter Landry <wlandry@caltech.edu>  Sun, 19 Feb 2017 12:00:00 -0700

tablator (1.0.1+dfsg-1) unstable; urgency=low

  * Read fits files that do not have the null_bitfield_flags column
  * Set the null value as 'nan' when reading fits files
  * Automatically expand arrays into columns when writing ipac_tables
  * Disable printing some HDF5 exceptions
  * Fix HTML output

 -- Walter Landry <wlandry@caltech.edu>  Mon, 13 Feb 2017 12:00:00 -0700

tablator (1.0.0+dfsg-1) unstable; urgency=low

  * Initial release (Closes: #843717)

 -- Walter Landry <wlandry@caltech.edu>  Sat, 05 Nov 2016 21:32:08 -0700


src/tablator.cxx
/// A simple converter program to test out the tablator library.

#include <CCfits/CCfits>
#include <boost/program_options.hpp>
#include <json5_parser.h>

#include "Table.hxx"

std::string usage (const boost::program_options::options_description
                   &visible_options)
{
  std::stringstream ss;
  ss << "Usage: tablator [options] input_file output_file\n"
     << visible_options;
  return ss.str ();
}

int main (int argc, char *argv[])
{
  bool stream_intermediate (false);
  tablator::Format input_format, output_format;

  // Declare the supported options.
  boost::program_options::options_description visible_options("Options");
  visible_options.add_options()
    ("help", "produce help message")
    ("stream-intermediate", boost::program_options::value<bool>(),
     "stream the intermediate file (for testing only)")
    ("input-format", boost::program_options::value<std::string>(),
     "Input file format (json,json5,votable,csv,tsv,fits,ipac_table,"
     "text,html,hdf5)")
    ("output-format", boost::program_options::value<std::string>(),
     "Output file format (json,json5,votable,csv,tsv,fits,ipac_table,"
     "text,html,hdf5)")
    ;

  boost::program_options::options_description
    hidden_options("Hidden options");
  hidden_options.add_options()
    ("files", boost::program_options::value<std::vector<std::string> >(),
     "input and output file")
    ;

  boost::program_options::options_description options;
  options.add (visible_options).add (hidden_options);
  boost::program_options::positional_options_description positional_options;
  positional_options.add("files", -1);

  boost::program_options::variables_map option_variables;
  try
    {
      boost::program_options::store
        (boost::program_options::command_line_parser
         (argc, argv).options(options).positional(positional_options).run(),
         option_variables);
      boost::program_options::notify(option_variables);

      if (option_variables.count("help"))
        {
          std::cout << usage(visible_options) << "\n";
          return 1;
        }
      if (option_variables.count("stream-intermediate"))
        stream_intermediate
          = option_variables["stream-intermediate"].as<bool>();
      if (option_variables.count("input-format"))
        input_format.init (option_variables["input-format"].as<std::string>());
      if (option_variables.count("output-format"))
        output_format.init (option_variables["output-format"].as<std::string>());

      boost::filesystem::path input_path, output_path;
      if (option_variables.count("files"))
        {
          std::vector<std::string> paths (option_variables["files"]
                                          .as<std::vector<std::string> >());
          if (paths.size()==1)
            {
              std::cerr << "Missing an output file\n";
              return 1;
            }
          if (paths.size()!=2)
            {
              std::cerr << "Too many filenames\n";
              return 1;
            }
          input_path=paths.at(0);
          output_path=paths.at(1);
          if (input_format.is_unknown ())
            input_format.set_from_extension(input_path);
          if (output_format.is_unknown ())
            output_format.set_from_extension(output_path);
        }
      else
        {
          std::cerr << "Missing an input and output file\n"
                    << usage (visible_options) << "\n";
          return 1;
        }

      H5::Exception::dontPrint ();
      tablator::Table table (input_path, input_format);

      if (stream_intermediate)
        {
          boost::filesystem::ofstream outfile (output_path);
          table.write_output (outfile, output_format);
        }
      else
        {
          table.write_output (output_path, output_format);
        }
    }
  catch (boost::program_options::error &exception)
    {
      std::cerr << exception.what () << "\n"
                << usage (visible_options);
      std::cerr.flush ();
      exit (1);
    }
  catch (CCfits::FitsException &exception)
    {
      std::cerr << exception.message () << "\n";
      std::cerr.flush ();
      exit (1);
    }
  catch (H5::Exception &exception)
    {
      std::cerr << "In " << exception.getFuncName () << ": "
                << exception.getDetailMsg () << "\n";
      std::cerr.flush ();
      exit (1);
    }
  catch (json5_parser::Error_position &exception)
    {
      std::cerr << "On line " << exception.line_ << ", column "
                << exception.column_ << ": " << exception.reason_ << "\n";
      std::cerr.flush ();
      exit (1);
    }
  catch (std::exception &exception)
    {
      std::cerr << exception.what () << "\n";
      std::cerr.flush ();
      exit (1);
    }
}
